core/domain/classifier_services.py

```diff
@@ -424,41 +392,212 @@ def delete_classifier_training_job(job_id):
+
+    training_job_exploration_mapping_model = (
+        classifier_models.TrainingJobExplorationMappingModel.get_model(
+            exp_id, exp_version, state_name))
+    training_job_exploration_mapping_model.algorithm_id_to_job_id_map = (
+        training_job_exploration_mapping.algorithm_id_to_job_id_map)
+    training_job_exploration_mapping_model.put()
+
+
+def get_classifier_training_job_maps(exp_id, exp_version, state_names):
+    """Gets the list of classifier training jobs mapped to algorithm id to from
+    the exploration attributes.
 
     Args:
         exp_id: str. ID of the exploration.
         exp_version: int. The exploration version.
         state_names: list(str). The state names for which we retrieve the job.
 
     Returns:
-        list(ClassifierTrainingJob). Domain objects for the Classifier training
-        job model.
+        list(dict(str: ClassifierTrainingJob)). Domain objects for the
+        Classifier training job model for algorithm_ids.
     """
     training_job_exploration_mapping_models = (
         classifier_models.TrainingJobExplorationMappingModel.get_models(
             exp_id, exp_version, state_names))
-    job_ids = []
+    algorithm_id_to_job_id_maps = []
     for mapping_model in training_job_exploration_mapping_models:
         if mapping_model is None:
             continue
-        job_ids.append(mapping_model.job_id)
+        algorithm_id_to_job_id_maps.append(
+            mapping_model.algorithm_id_to_job_id_map)
+
+    job_id_to_algorithm_id_mapping = {
+        job_id: {
+            'algorithm_id': algorithm_id,
+            'state_name': mapping_model.state_name
+        } for mapping_model, map in python_utils.ZIP(
+            training_job_exploration_mapping_models,
+            algorithm_id_to_job_id_maps)
+        for algorithm_id, job_id in map.items()
+    }
+    job_ids = job_id_to_algorithm_id_mapping.keys()
 
     classifier_training_job_models = (
         classifier_models.ClassifierTrainingJobModel.get_multi(job_ids))
 
-    classifier_training_jobs = []
+    state_to_algorithm_id_job_id_map = {}
     for job_model in classifier_training_job_models:
-        classifier_training_jobs.append(get_classifier_training_job_from_model(
-            job_model))
-
-    # Backfill None's to maintain indexes.
-    for index, mapping_model in enumerate(
-            training_job_exploration_mapping_models):
-        if mapping_model is None:
-            classifier_training_jobs.insert(index, None)
-    return classifier_training_jobs
+        job_id = job_model.id
+        algorithm_id = job_id_to_algorithm_id_mapping[job_id]['algorithm_id']
+        state_name = job_id_to_algorithm_id_mapping[job_id]['state_name']
+        if state_name not in state_to_algorithm_id_job_id_map:
+            state_to_algorithm_id_job_id_map[state_name] = {}
+        state_to_algorithm_id_job_id_map[state_name][algorithm_id] = job_model
+
+    algorithm_id_to_training_job_maps = []
```

Comment:
This doesn't look like a map to me? (looks more like list(str))?

Reply:
Its a list of maps. A list(dict).